### ISSUE TITLE
Support JSON path

### ISSUE THREAD

**Issue by Sir4ur0n** [yesodweb/persistent#1021]
Hi,

Sorry if this is a duplicate, I did not find any issue about it.
Does Persistent support [JSON path](https://www.postgresql.org/docs/12/datatype-json.html#DATATYPE-JSONPATH) syntax?

I know that e.g. `persistent-postgresql` supports JSONB PostgreSQL-specific operators like `@>` but I found no occurrence of jsonpath operators like `@@` or `@?`.

And if it doesn't: what do you think about adding it?

Cheers

**Comment by parsonsmatt** [comment]
I'd be happy to accept a patch to `persistent-postgresql` that adds support for this explicitly. I'm curious what API you'd like to have and what use cases you want to support. 

Looking at the `@?` and `@@` operators, I think we can provide direct support for that using a custom `BackendSpecificFilter` and `Filter` values. For a first pass, you could write something like:

```haskell
(@?) 
    :: (PersistEntity entity)
    => EntityField entity Value
    -> Text -- ^ The JSONPath string
    -> Filter entity
entityField @? jsonPath =
  Filter
    { filterField = entityField
    , filterValue = UnsafeValue jsonPath
    , filterFilter = BackendSpecificFilter jsonPath
    }
```

You can use this like:

```haskell
mkPersist sqlSettings [persistLowerCase|
User
  custom Value
|]

selectList [UserCustom @? "$.track.segments[*].HR < 70"] []
```

(stealing the query from [the postgres docs](https://www.postgresql.org/docs/12/functions-json.html#FUNCTIONS-SQLJSON-PATH))

That would generate a query something like:

```sql
SELECT *
FROM user
WHERE user.custom @? '$.track.segments[*].HR < 70';
```

**Comment by Vlix** [comment]
That would work, but does feel kind of unsafe (while to me, `persistent` is like the "easy and safe" database library). This feels like something that would fit better in the `esqueleto` package, IMHO.
Or maybe if you can make a nice API to construct a `JSONPath`? 🤔

**Comment by parsonsmatt** [comment]
Yeah you'd definitely want to come up with a safe API for this - the strings would Work, but only in the `persistent` sense of "If we don't support it, you can a) build a library on top that does (eg esqueleto) or b) use a raw string query".
